components/brave_wallet/browser/solana_data_decoder_utils.cc

```diff
@@ -0,0 +1,204 @@
+#include "components/grit/brave_components_strings.h"
+#include "ui/base/l10n/l10n_util.h"
+
+namespace brave_wallet {
+
+constexpr uint8_t kAuthorityTypeMax = 3;
+constexpr size_t kMaxStringSize32Bit = 4294967291u;
+
+absl::optional<uint8_t> DecodeUint8(const std::vector<uint8_t>& input,
+                                    size_t& offset) {
+  if (offset >= input.size() || input.size() - offset < sizeof(uint8_t)) {
+    return absl::nullopt;
+  }
+
+  offset += sizeof(uint8_t);
+  return input[offset - sizeof(uint8_t)];
+}
+
+absl::optional<std::string> DecodeUint8String(const std::vector<uint8_t>& input,
+                                              size_t& offset) {
+  auto ret = DecodeUint8(input, offset);
+  if (!ret)
+    return absl::nullopt;
+  return base::NumberToString(*ret);
+}
+
+absl::optional<std::string> DecodeAuthorityTypeString(
+    const std::vector<uint8_t>& input,
+    size_t& offset) {
+  auto ret = DecodeUint8(input, offset);
+  if (ret && *ret <= kAuthorityTypeMax)
+    return base::NumberToString(*ret);
+  return absl::nullopt;
+}
+
+absl::optional<uint32_t> DecodeUint32(const std::vector<uint8_t>& input,
+                                      size_t& offset) {
+  if (offset >= input.size() || input.size() - offset < sizeof(uint32_t)) {
+    return absl::nullopt;
+  }
+
+  // Read bytes in little endian order.
+  base::span<const uint8_t> s =
+      base::make_span(input.begin() + offset, sizeof(uint32_t));
+  uint32_t uint32_le = *reinterpret_cast<const uint32_t*>(s.data());
+
+  offset += sizeof(uint32_t);
+
+#if defined(ARCH_CPU_LITTLE_ENDIAN)
+  return uint32_le;
+#else
+  return base::ByteSwap(uint32_le);
+#endif
+}
+
+absl::optional<std::string> DecodeUint32String(
+    const std::vector<uint8_t>& input,
+    size_t& offset) {
+  auto ret = DecodeUint32(input, offset);
+  if (!ret)
+    return absl::nullopt;
+  return base::NumberToString(*ret);
+}
+
+absl::optional<uint64_t> DecodeUint64(const std::vector<uint8_t>& input,
+                                      size_t& offset) {
+  if (offset >= input.size() || input.size() - offset < sizeof(uint64_t)) {
+    return absl::nullopt;
+  }
+
+  // Read bytes in little endian order.
+  base::span<const uint8_t> s =
+      base::make_span(input.begin() + offset, sizeof(uint64_t));
+  uint64_t uint64_le = *reinterpret_cast<const uint64_t*>(s.data());
+
+  offset += sizeof(uint64_t);
+
+#if defined(ARCH_CPU_LITTLE_ENDIAN)
+  return uint64_le;
+#else
+  return base::ByteSwap(uint64_le);
+#endif
+}
+
+absl::optional<std::string> DecodeUint64String(
+    const std::vector<uint8_t>& input,
+    size_t& offset) {
+  auto ret = DecodeUint64(input, offset);
+  if (!ret)
+    return absl::nullopt;
+  return base::NumberToString(*ret);
+}
+
+absl::optional<std::string> DecodePublicKey(const std::vector<uint8_t>& input,
+                                            size_t& offset) {
+  if (offset >= input.size() || input.size() - offset < kSolanaPubkeySize)
```

Comment:
This one as well for separate thread to parse it.

Also shouldn't this be: `offset >= input.size() || input.size() - offset != kSolanaPubkeySize` since the key size would be a set value

Reply:
No, input array doesn't only contain public key, public key is just a part of the data input.